src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -19627,49 +19785,311 @@ class C
+                //         var a3 = new[] { x, y! }; // 3
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "y!").WithArguments("C<object?>", "C<object>").WithLocation(8, 29)
+                );
+        }
+
+        [Fact]
+        public void SuppressNullableWarning_LocalDeclaration()
+        {
+            var source =
+@"
+class C<T>
+{
+    static void F(C<object>? x, C<object?> y)
+    {
+        C<object>? c1 = y; // 1
+        C<object?> c2 = x; // 2 and 3
+        C<object>? c3 = y!; // 4
+        C<object?> c4 = x!; // 5
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (6,25): warning CS8619: Nullability of reference types in value of type 'C<object?>' doesn't match target type 'C<object>'.
+                //         C<object>? c1 = y; // 1
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "y").WithArguments("C<object?>", "C<object>").WithLocation(6, 25),
+                // (7,25): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         C<object?> c2 = x; // 2 and 3
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "x").WithLocation(7, 25),
+                // (7,25): warning CS8619: Nullability of reference types in value of type 'C<object>' doesn't match target type 'C<object?>'.
+                //         C<object?> c2 = x; // 2 and 3
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "x").WithArguments("C<object>", "C<object?>").WithLocation(7, 25),
+                // (8,25): warning CS8619: Nullability of reference types in value of type 'C<object?>' doesn't match target type 'C<object>'.
+                //         C<object>? c3 = y!; // 4
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "y!").WithArguments("C<object?>", "C<object>").WithLocation(8, 25),
+                // (9,25): warning CS8619: Nullability of reference types in value of type 'C<object>' doesn't match target type 'C<object?>'.
+                //         C<object?> c4 = x!; // 5
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "x!").WithArguments("C<object>", "C<object?>").WithLocation(9, 25)
+                );
+        }
+
+        [Fact]
+        public void SuppressNullableWarning_Cast()
+        {
+            var source =
+@"
+class C<T>
+{
+    static void F(C<object>? x, C<object?> y)
+    {
+        var c1 = (C<object>?)y;
+        var c2 = (C<object?>)x; // warn
+        var c3 = (C<object>?)y!;
+        var c4 = (C<object?>)x!;
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            // PROTOTYPE(NullableReferenceTypes): Should there be a warning on c1 too?
```

Comment:
No warning on `c1` since the explicit cast changes nested nullability.

Reply:
My understanding of cast is that it does not silently change nested nullability.
If that's ok, I'll leave this comment until we review the rules for cast in LDM.

---
In reply to: [194186240](https://github.com/dotnet/roslyn/pull/27317#discussion_r194186240) [](ancestors = 194186240)